vimeo/vimeo-networking-java
# Folders CRUD support

# Summary

Supporting CRUD for folders. Adds the ability to create folders, edit folder settings, add videos to folders, removing videos from folders, and deleting folders. Additionally, I added the ability to get a list of folders, which required creating a `FolderList` DTO. The functions added are
- `createFolder(uri)`
- `createFolder(user)`
- `editFolder(uri)`
- `editFolder(folder)`
- `addToFolder(folder, video)`
- `addToFolder(folderUri, videoUri)`
- `removeFromFolder(folder, video)`
- `removeFromFolder(folderUri, videoUri)`
- `fetchFolderList(uri)`
- `deleteFolder(folder)` - no URI equivalent exists for this since `deleteContent(uri)` covers that use case.

I removed the following function which was duplicated by `deleteContent(uri)`
- `deleteAlbum(uri)`

I also added a new connection to `UserConnections` which returns all the folders of the user. I called it `folders` since it returns all the user's folders. Since we already had a `folders` connection which returned the projects in the root directory, I renamed that to `projectItemsRoot`. I think this is more clear than before, where the `folders` connection returned `ProjectItemList` instead of `FolderList`. Now `folders` returns a `FolderList` and the `projectItemsRoot` returns a `ProjectItemList`.

Additionally, I added the `parentFolder` property to a `Video` which allows us to see what folder (if any) that a video is currently in.

## Other
I fixed a bug where the consumer could try to clear empty URIs which results in a crash.

## Testing
1. Log in with `Authenticator.authenticateWithEmailLogin`
2. Create a folder by passing `Authenticator.currentAccount?.user` to `VimeoApiClient.createFolder(user, name, privacy)`
3. Fetch the current folders by passing `Authenticator.currentAccount?.user?.metadata?.connections?.folders?.uri` to `VimeoApiClient.fetchFolderList(uri)` to verify that the folder was created.
4. Add a video to the folder by passing the previously created folder and a video uploaded by the user to `VimeoApiClient.addToFolder(folder, video)`.
5. Remove the video from the folder by passing the previously created folder and a video uploaded by the user to `VimeoApiClient.removeFromFolder(folder, video)`.
6. Delete the folder by passing the folder previously created to `VimeoApiClient.deleteFolder(folder)`.
